lib/schema_matching.js

```diff
@@ -0,0 +1,331 @@
+    matched_headers = []
+    for (const header of Object.keys(crosswalk_data)) {
+        if (crosswalk_data[header].includes(colname.toLowerCase())) {
+            matched_headers.push(header)
+        }
+    }
+    if (matched_headers.length > 0) {
+        return matched_headers
+    } else {
+        return 'There is no matched history for the column name "' + colname + '"'
+    }
+}
+
+/**
+ * This function takes in a column name and finds the target headers that might be associated with the column name,
+ * based on patterns in the crosswalk data.
+ * 
+ * @param {string} colname from a dataset
+ * @param {object} crosswalk_data sortOriginalHeadersByTargetHeader(removeUnitSuffixesAndFunctions(getCrosswalkData()))
+ * @returns {Array|string} either an array with the similar headers or a string stating there are no similar headers
+ */
+function checkSimilarHeaders(colname, crosswalk_data) {
+    similar_headers = []
+    // patterns found in the history of matching for each target header
+    potential_similar_headers = {
+        'scientific':['botanical'],
+        'ref':['id'],
+        'health':['condition'],
+        'updated':['edit','date'],
+        'dbh':['diameter'],
+        'crown':['spread'],
+        'note':['comment'],
+        'cultivar':['variete'],
+        'description':['desc'],
+        'age':['treeage'],
+    }
+    for (const header of Object.keys(crosswalk_data)) {
+        if (colname.toLowerCase().includes(header)) {
+            similar_headers.push(header)
+        }
+    }
+    for (const header of Object.keys(potential_similar_headers)) {
+        if (colname.toLowerCase().includes(potential_similar_headers[header])) {
+            if (!similar_headers.includes(header)) {
+                similar_headers.push(header)
+            }
+        }
+    }
+    for (const header_pairs of Object.entries(crosswalk_data)) {
+        for (const matched_header of header_pairs[1]) {
+            if (typeof matched_header == 'string') {
+                if (colname.toLowerCase().includes(matched_header.toLowerCase())) {
+                    if (!similar_headers.includes(header_pairs[0])) {
+                        similar_headers.push(header_pairs[0])
+                    }     
+                }
+            }
+        }
+    }
+    if (similar_headers.length > 0) {
+        return similar_headers
+    } else {
+        return 'There are no guesses for the column name "' + colname + '"'
+    }
+}
+
+/**
+ * This function takes in information from a dataset and the crosswalk data, and provides a guess on the crosswalk for the dataset.
+ * The user should copy the output and manually edit it.
+ * 
+ * @param {Array} headers from a dataset
+ * @param {object} crosswalk_dataset sortOriginalHeadersByTargetHeader(removeUnitSuffixesAndFunctions(getCrosswalkData()))
+ */
+function makeCrosswalkMapping(headers, crosswalk_dataset) {
+    crosswalk = {}
+    ignored_columns = []
+    ignored_bc_existing = {}
```

Comment:
Why do we have ignored and ignored bc existing?

Reply:
ignored_columns are colmuns that the guesser didn't find a guess for (so they're ignored). ignored_bc_existing are columns where the guesser's guess already exists in the crosswalk (so they're ignored).